File: src/56_01_NumbersAppearOnce/56_01_NumbersAppearOnce.cpp
// 面试题56（一）：数组中只出现一次的两个数字
// 题目：一个整型数组里除了两个数字之外，其他的数字都出现了两次。请写程序
// 找出这两个只出现一次的数字。要求时间复杂度是O(n)，空间复杂度是O(1)。

#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
   public:
    // 利用异或
    vector<int> singleNumbers( vector<int>& nums ) {
        if ( nums.size() < 2 ) return {};

        // 由于数组里除了两个数字之外,其余只出现了两次
        // 有 a [xor] a ... [xor] x [xor] y = x [xor] y
        int n = 0;
        for ( int num : nums ) n ^= num;  // n = x [xor] y

        if ( n == 0 ) return {};  // 处理无单独数字的情况

        // 接下来从n中拆分x和y. 若x xor y的某位为1, 说明xy该位不同.
        // 找到某个为1的二进制位, 据此将数组拆分为两组.
        // 再次遍历数组, 按照该位的分类对x或y进行异或, 遍历完所有数字即得到x和y

        int index = 1;
        while ( ( n & index ) == 0 ) index <<= 1;  // 找到n中为1的最低位

        int x = 0, y = 0;
        for ( int num : nums ) {  // 分组遍历
            if ( num & index )
                x ^= num;
            else
                y ^= num;
        }
        return { x, y };
    }
};

class Test {
   public:
    void checkResult( const string testName, vector<int>& nums, int expected1,
                      int expected2, bool valid ) {
        cout << testName;
        vector<int> results = solution.singleNumbers( nums );
        if ( results.size() == 2 && results[ 0 ] != results[ 1 ] && valid ) {
            bool find1 = std::find( results.begin(), results.end(),
                                    expected1 ) != results.end();
            bool find2 = std::find( results.begin(), results.end(),
                                    expected2 ) != results.end();
            if ( find1 && find2 )
                cout << " passed.\n";
            else
                cout << " failed.\n";
        } else if ( ( results.size() < 2 || results[ 0 ] == results[ 1 ] ) &&
                    !valid )
            cout << " passed.\n";
        else
            cout << " failed.\n";
    }

    void Test1() {
        vector<int> nums{ 2, 4, 3, 6, 3, 2, 5, 5 };
        checkResult( "Test1", nums, 4, 6, true );
    }

    void Test2() {
        vector<int> nums{ 4, 6 };
        checkResult( "Test2", nums, 4, 6, true );
    }

    void Test3() {
        vector<int> nums{ 4, 6, 1, 1 };
        checkResult( "Test3", nums, 4, 6, true );
    }

    void Test4() {
        vector<int> nums{ 4 };
        checkResult( "Test4", nums, -1, -1, false );
    }

    void Test5() {
        vector<int> nums{ 1, 1, 1, 1 };
        checkResult( "Test5", nums, -1, -1, false );
    }

    void Test6() {
        vector<int> nums;
        checkResult( "Test6", nums, -1, -1, false );
    }

    void runTest() {
        Test1();
        Test2();
        Test3();
        Test4();
        Test5();
        Test6();
    }

   private:
    Solution solution;
};

int main() {
    Test test;
    test.runTest();
    return 0;
}

File: src/56_01_NumbersAppearOnce/56_01_NumbersAppearOnce.py
class Solution:
    def singleNumbers(self, nums: list[int]) -> list[int]:
        if len(nums) < 2:
            return []

        n = 0
        for num in nums:
            n ^= num

        if n == 0:
            return []

        index = 1
        while (n & index) == 0:
            index <<= 1

        x, y = 0, 0
        for num in nums:
            if num & index:
                x ^= num
            else:
                y ^= num
        return [x, y]


class Test:
    def __init__(self) -> None:
        self.solution = Solution()

    def checkResult(
        self,
        testName: str,
        nums: list[int],
        expected1: int,
        expected2: int,
        valid: bool,
    ) -> None:
        text = testName
        results = self.solution.singleNumbers(nums)
        if len(results) == 2 and results[0] != results[1] and valid:
            if expected1 in results and expected2 in results:
                text += " passed."
            else:
                text += " failed."
        elif (len(results) < 2 or results[0] == results[1]) and not valid:
            text += " passed."
        else:
            text += " failed."
        print(text)

    def Test1(self):
        nums = [2, 4, 3, 6, 3, 2, 5, 5]
        self.checkResult("Test1", nums, 4, 6, True)

    def Test2(self):
        nums = [4, 6]
        self.checkResult("Test2", nums, 4, 6, True)

    def Test3(self):
        nums = [4, 6, 1, 1]
        self.checkResult("Test3", nums, 4, 6, True)

    def Test4(self):
        nums = [4]
        self.checkResult("Test4", nums, -1, -1, False)

    def Test5(self):
        nums = [1, 1, 1, 1]
        self.checkResult("Test5", nums, -1, -1, False)

    def Test6(self):
        nums = []
        self.checkResult("Test6", nums, -1, -1, False)

    def runTest(self):
        self.Test1()
        self.Test2()
        self.Test3()
        self.Test4()
        self.Test5()
        self.Test6()


if __name__ == "__main__":
    test = Test()
    test.runTest()
